Properly patch statefulset

alenkacz opened the pull request:
## Cover letter

Currently we update only replicas, resources and image. This is problematic
because e.g. external connectivity and other features cannot get enabled as
well as disabled, because those fields don't get updated.

This PR introduces k8s-objectmatcher library that will help us perform the
patches of the statefulset and keep everything up to date.

If we agree on this, we will similarly patch also all other objects as a follow-up.

Fixes issues: partially addresses #811

## Release notes

If the PR changes the user experience, write a short summary of the changes. See the [CONTRIBUTING](https://github.com/vectorizedio/redpanda/blob/dev/CONTRIBUTING.md) guidelines for details.

Release note: [1-2 sentences of what this PR changes]
Statefulset will now be updated when cluster CR is updated

RafalKorepta commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -83,12 +81,6 @@ func (r *Cluster) ValidateUpdate(old runtime.Object) error {
 				r.Spec.Replicas,
 				"scaling down is not supported"))
 	}
-	if !reflect.DeepEqual(r.Spec.Configuration, oldCluster.Spec.Configuration) {
```
With this change the test description need to be updated

Remove: 

> (unable to update configuration)

- https://github.com/vectorizedio/redpanda/pull/822/commits/4de256127edbd25ff462ac4caf19f03a07f76352#diff-46bd3dd552497ca6d98b48ef0c41da598b2090731a86a4880c46d94d76c5bb7dR121
- https://github.com/vectorizedio/redpanda/pull/822/commits/4de256127edbd25ff462ac4caf19f03a07f76352#diff-46bd3dd552497ca6d98b48ef0c41da598b2090731a86a4880c46d94d76c5bb7dR131
- https://github.com/vectorizedio/redpanda/pull/822/commits/4de256127edbd25ff462ac4caf19f03a07f76352#diff-46bd3dd552497ca6d98b48ef0c41da598b2090731a86a4880c46d94d76c5bb7dL142

alenkacz replied:
oh okay, I missed that, thanks

RafalKorepta replied:
The tests should be ok, but the description is off.

RafalKorepta reviewed (commented):
> operator: e2e test enabling tls
>
> This is something that would be failing on dev because we never patch those fields.

I don't understand the context provided. Could you describe the test added? What value it brings? 

Hypothetically if this commit land it should not describe what would happen in `dev` as it changing frequently :)

alenkacz commented:
Updated. I also updated the commit description. 

The original story was trying to communicate, that currently what the test is doing won't work on dev, so it's something this PR actually fixes. We were not allowing people to patch the configurarion so that was not possible up until now. Apart from that the only fields we were updating were replicas, image and resources.

dimitriscruz reviewed (changes requested):
Where do we set the "upgrading" status to true? (it was removed)

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -44,32 +44,13 @@ const requeueDuration = time.Second * 10
-
-	if upgrading {
-		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
+	if r.pandaCluster.Status.Upgrading {
```
thanks! 🤦‍♀️

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -44,32 +44,13 @@ const requeueDuration = time.Second * 10
 // verification checks the pod has started listening in its Kafka API port and may be
 // extended.
-func (r *StatefulSetResource) updateStsImage(
+func (r *StatefulSetResource) runPartitionedUpdate(
```
Yeah good point, anyway I just chatted with Michal and I think we should be using this for any update that would restart container, I'll be changing this in some upcoming PR. Are you fine with leaving this as it is right now even though right now it's still handling only image?

alenkacz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -134,58 +140,55 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
 	}
-
-	return false
+	if !patchResult.IsEmpty() {
```
I've extracted the whole update method to resource.go https://github.com/vectorizedio/redpanda/pull/822/commits/e22531e5c8ace31bd55e630246747879ae4b58ba let me know if that works for you

RafalKorepta reviewed (approved):
LGTM
